Fix regression in allowed_attributes and allowed_interface

pbauer opened the pull request:
See https://github.com/zopefoundation/Zope/pull/643

MatthewWilkes committed:
Attempted fix for allowed_attributes regression.

The simple view mixin caused a regression on allowed_attributes and allowed_interface. This fixes that by explicitly whitelisting those attributes for traversal.

MatthewWilkes committed:
Explicitly rebind simple.__call__ when overriding the method for Python 2 support

mgedmin commented on src/zope/browserpage/metaconfigure.py:
```diff
@@ -411,7 +412,12 @@ def _handle_for(_context, for_):
-        raise NotFound(self, name, request)
+        if name in getattr(self, "_simple_whitelist", []):
+            self.__page_attribute__ = name
+            self.__call__ = simple.__call__.__get__(self, simple)
```
What is this line doing and why?

MatthewWilkes replied:
`simple` is added as a base alongside whatever class the user defined. If the user defined `__call__` then this will win, over the one in `simple`. We need the `__call__` attribute to point to the way of rendering the view, so the version in `simple` actually is perfect for what we need. In this case, we want to make the instance use the `simple.__call__` implementation rather than the one the class alone would pick.

The way I do this is by setting `__call__` on the instance to point at the right function. In Python3 we can just do `self.__call__ = simple.__call__` because in Python3 `type(simple.__call__)` is `<class 'function'>`. In Python2 it's `<type 'instancemethod'>`. This is to do with the whole new/old style class compatibility thing back in Python2.

This line uses the fact that instancemethods and functions are descriptors to explicitly bind it to the instance when setting it on the instance dictionary. If we don't do this then `self` looks like just another parameter to Python2 as it's not gone through the normal method resolution protocol.

mgedmin replied:
Would `self.__call__ = self.__call__`  work to invoke the descriptor protocol?

But I guess my main question was about why setting `__call__` on an instance even works?  Is the publisher doing a `getattr(obj, '__call__')` explicitly?  Because Python itself won't use `__call__` of an instance, it'll only look up special methods from the class.

mgedmin replied:
BTW all the tests pass if I comment the `self.__call__` assignment, so either it's not necessary, or the tests are inadequate.

MatthewWilkes replied:
Maybe `self.__call__ = self.__call__` would work, I'm not sure. It looks even weirder to me, though.

Yes, it works because there's an explicit getattr for call. This is actually done in code that's relevant to this PR, in https://github.com/zopefoundation/zope.browserpage/blob/master/src/zope/browserpage/metaconfigure.py#L146 so we could make this a lower impact change by changing the relevant method to a dispatcher and allowing it to call the discovered `__call__`.

Yeah, I've only added pretty minimal tests. To be honest, this is a really busy time for me and I jumped in on this because I was directly emailed a couple of times to ask for my opinion. I meant this more as a demonstration of the approach I would take than a production-ready PR. I'm happy to update it but equally I'm happy for people to just change the branches or branch themselves to get something they're confident in.

icemac replied:
Could it be that this assignment is only necessary if a child class has its own `__call__` to reset it to actually get the attribute? Maybe there should be a test demonstrating its necessity.

mgedmin commented on src/zope/browserpage/metaconfigure.py:
```diff
@@ -168,6 +168,7 @@ def page(_context, name, permission, for_=Interface,
                                required)
 
     _handle_for(_context, for_)
+    new_class._simple_whitelist = set(required) - set([attribute, 'browserDefault', '__call__', 'publishTraverse'])
```
Maybe this should be `_simple__whitelist` with the double underscore that you'd get if you used `self.__simple` inside the `simple` class?  To prevent clashes with arbitrary subclasses etc.

MatthewWilkes replied:
Probably a good idea.

pbauer replied:
done in https://github.com/zopefoundation/zope.browserpage/commit/dcdda5f3fa0ad4f9f79ce0cf593c33f548f083b5 and https://github.com/zopefoundation/Zope/pull/643/commits/6c1134448fed1bfc25076f3738c977b541ce8ce4

mgedmin commented:
This PR makes `tox -e coverage` fail because our tox.ini has `--fail-under=100`.

Apparently four lines inside .../zope/browserpage/tests/test_metaconfigure.py are not covered: lines 45, 98, 114, and 119.

I propose `# pragma: no cover` on the first three, and replacing `pass` with a `"foo"` docstring for line 119.

mgedmin commented:
(I also propose adding `-m` to `coverage report` in tox.ini)

d-maurer commented:
Matthew Wilkes wrote at 2019-5-28 10:43 -0700:
>MatthewWilkes commented on this pull request.
>
>
>
>> @@ -411,7 +412,12 @@ def _handle_for(_context, for_):
> class simple(BrowserView):
>
>     def publishTraverse(self, request, name):
>-        raise NotFound(self, name, request)
>+        if name in getattr(self, "_simple_whitelist", []):
>+            self.__page_attribute__ = name
>+            self.__call__ = simple.__call__.__get__(self, simple)
>
>Maybe `self.__call__ = self.__call__` would work, I'm not sure. It looks even weirder to me, though.

Usually, an instance level definition of a special method
(often indicated by the form "__XXXX__") are ignored.
Hopefully, this is not the case of `__call__`.

MatthewWilkes committed:
Simplify allowed_attributes path

MatthewWilkes committed:
Simplify allowed_attributes path

mgedmin reviewed (approved):
This looks much simpler, thanks!

(I wonder if there are any backwards-compatibility concerns about `__call__` going away.  I guess we'll find out once this is merged and released!)

mgedmin commented:
Ah, even zope.browserpage's own testsuite wants instances of `simple` to be callable. :/

mgedmin committed:
Merge branch 'master' into wilkes/397-allowed-attributes

mgedmin commented:
(Apologies for introducing merge conflicts with my Python 3.4 drop.  I've resolved them, hope that was ok!)

MatthewWilkes committed:
Merge branch 'wilkes/397-allowed-attributes' of https://github.com/zopefoundation/zope.browserpage into wilkes/397-allowed-attributes

MatthewWilkes commented:
@mgedmin Hey, sorry, I was still playing with this in the background, hadn't seen you were working on it too. It's hard trying to run the test suite of all these things at once, I was focused on the Plone side and hadn't looked at Zope yet. I've readded `__call__`, let's see how this all looks.

MatthewWilkes commented:
Okay, there are failing tests in here. `testNamedViewNoPagesForNonCallable` seems to be the main one, but I'm not sure it's testing the right thing. I've got to leave this for today (and I likely won't have any time tomorrow), but I think we're close.

The Plone test suite is failing ( https://jenkins.plone.org/job/pull-request-5.2-3.7/217/testReport/ ) but it looks like the failing tests are all in `zope.browserpage` so I think we're close.

pbauer committed:
I'm just guessing here...

pbauer committed:
rename _simple_whitelist to _simple__whitelist

pbauer commented:
I restarted the tests with my changes: https://jenkins.plone.org/job/pull-request-5.2-3.7

Commit:
Bring coverage back to 100 %

Commit:
Provide a default `__page_attribute__`.

This way `browserDefault` does not die with an AttributeError but
`__call__` will do it.

icemac commented:
I added a default for `__page_attribute__` as this seems more natural for me. This makes that `browserDefault` does not break but delegates this to `__call__` if there is no custom one.
I am not sure if this implementation is correct.

@MatthewWilkes @mgedmin Any thoughts?
@pbauer Could you please try the Plone tests against my changes?

icemac commented on src/zope/browserpage/metaconfigure.py:
```diff
@@ -410,20 +404,36 @@ def _handle_for(_context, for_):
+            raise NotFound(self, name, request)
 
     def __call__(self, *a, **k):
         # If a class doesn't provide it's own call, then get the attribute
```
Maybe this line should be removed, too.

icemac replied:
The same way it was done with the following line (see diff.)

pbauer commented:
Looks good locally and I started the tests for https://github.com/plone/buildout.coredev/pull/588

pbauer commented:
@icemac all test are green.
